### DISCUSSION TITLE
Port System.Printing to C#

### DISCUSSION THREAD

**Discussion by bgrainger** [dotnet/wpf#6663] [General]
System.Printing is currently a C++/CLI assembly. There would be advantages to rewriting it in C#, as detailed in this comment: https://github.com/dotnet/wpf/issues/5305#issuecomment-922067322 CC @ThomasGoulet73

I haven't reviewed all the code in the project line-by-line. There is obviously some P/Invoke, some COM, and some native types in `AsyncNotifyUnmanaged.cpp`, which should make the port potentially complex, but not impossible. 

Are there critical blocking issues that would make a port to C# impossible? What are the reference cycles related to System.Printing and would those still exist following a C# port?

Or is this primarily about the effort of rewriting thousands of lines of C++/CLI in C#?

**Comment by kant2002** [comment]
https://github.com/kant2002/wpf/commit/d91ae727247f225c827654ffb9cbc5edea411b72

I done that using ILSpy and it requires A LOT OF polish, but my experience in doing that tells me that first should be solved cyclic dependencies in non-breaking fashion and would be good if that would would be accepted here. We should try maintain compatibility with .NET Framework, otherwise whole work would be futile IMO. So I think in order to successfully port `System.Printing` to C# would be good to solve cyclic dependencies first, and that proceed with rewrite. That would be massively easier for anybody.

Effort for me was not doing that in one big rewrite attempt, like it was with `DirectWriteForwarder` and I cannot find a way to do that, if somebody can give me ideas, I can try doing porting in more reliable fashion.

Porting that code would be **very mundane**, but I would say relatively easy, since nothing depends on unmanaged parts, like TTF reading in `DirectWriteForwarder`.

**Reply by huoyaoyuan** [reply]
What's the cyclic dependency problem here?

**Reply by lindexi** [reply]
@huoyaoyuan I do not sure what the cyclic dependency problem he means. But the wpf need the `cycle-breakers` projects to fix the cyclic dependency issues, see https://github.com/dotnet/wpf/tree/main/src/Microsoft.DotNet.Wpf/cycle-breakers/System.Printing

**Comment by wstaelens** [comment]
In case somebody works in printing, pretty please take a look at:

- https://github.com/dotnet/wpf/issues/3546
- https://github.com/dotnet/wpf/issues/4000
- https://github.com/dotnet/wpf/issues/6296 (and the list of XPS/printing issues in this topic)

**Comment by batzen** [comment]
@wstaelens I don't think that porting to C# should be mixed with bug fixes.

**Comment by wstaelens** [comment]
@batzen regarding #3546 is an issues as a result from porting .NET Framework to .NET (Core) where a part was not converted (properly). The standard microsoft print driver from the DDK generates XPS files when printing, but these XPS files generated from the default print driver in the DDK cannot be opened because something was not fully ported. The XPS is generated in .piece files but that part was forgotten.

I don't know how deeply it is nested in System.Printing or if it is only in the System.IO.Packaging but anyway, if somebody is looking to porting System.Printing, this should be looked at as it involves printing and it is broken now.
I just don't hope somebody ports code from .NET to fully managed C#, while the .NET port from .NET Framework itself isn't ok.
Hence my comment. 

The other onces are mainly memory issues that could be maybe easily resolved when code runs in C#. Fixing memory issues in managed code is often easier. Thus important to check when porting imo.

**Comment by bgrainger** [comment]
FWIW, in my mind this would be a line-by-line port (by hand) of each file, preserving the existing comments and semantics exactly as they are in the C++/CLI. (Any deviation from the C++/CLI behaviour in the port would be considered a bug.)
This would provide a base upon which bug fixes, optimisations, etc. could be made.

**Comment by kant2002** [comment]
Would be good if some test application/test was written or collected, because that rewrite should be tested not only manually, but also on working code. That will helps a lot with `surpise issues` later on.

**Comment by wstaelens** [comment]
@bgrainger what is the progress of the port to c# ?

**Reply by bgrainger** [reply]
No one has started it AFAIK. Are you volunteering to lead the effort?

**Reply by lindexi** [reply]
I have no plans for this year...
